src/coreclr/jit/gentree.h

```diff
@@ -5116,18 +5049,177 @@ struct GenTreeIntrinsic : public GenTreeOp
+// utilizing GenTreeMultiOp. GTF_REVERSE_OPS is supported for GenTreeMultiOp's
+// with two operands.
+//
+struct GenTreeMultiOp : public GenTree
 {
+public:
+    class Iterator
+    {
+    protected:
+        GenTree** m_use;
+
+        Iterator(GenTree** use) : m_use(use)
+        {
+        }
+
+    public:
+        Iterator& operator++()
+        {
+            m_use++;
+            return *this;
+        }
+
+        bool operator==(const Iterator& other) const
+        {
+            return m_use == other.m_use;
+        }
+
+        bool operator!=(const Iterator& other) const
+        {
+            return m_use != other.m_use;
+        }
+    };
+
+    class OperandsIterator final : public Iterator
+    {
+    public:
+        OperandsIterator(GenTree** use) : Iterator(use)
+        {
+        }
+
+        GenTree* operator*()
+        {
+            return *m_use;
+        }
+    };
+
+    class UseEdgesIterator final : public Iterator
+    {
+    public:
+        UseEdgesIterator(GenTree** use) : Iterator(use)
+        {
+        }
+
+        GenTree** operator*()
+        {
+            return m_use;
+        }
+    };
+
 private:
-    ClassLayout* gtLayout;
+    GenTree** m_operands;
 
-    unsigned char  gtAuxiliaryJitType; // For intrinsics than need another type (e.g. Avx2.Gather* or SIMD (by element))
-    regNumberSmall gtOtherReg;         // For intrinsics that return 2 registers
+protected:
+    template <unsigned InlineOperandCount, typename... Operands>
+    GenTreeMultiOp(genTreeOps    oper,
+                   var_types     type,
+                   CompAllocator allocator,
+                   GenTree* (&inlineOperands)[InlineOperandCount] DEBUGARG(bool largeNode),
+                   Operands... operands)
+        : GenTree(oper, type DEBUGARG(largeNode))
+    {
+        const size_t OperandCount = sizeof...(Operands);
+
+        // "OperandCount + 1" so that it works well when OperandCount is 0.
+        GenTree* operandsArray[OperandCount + 1]{operands...};
+        InitializeOperands(allocator, operandsArray, OperandCount, inlineOperands, InlineOperandCount);
+    }
 
-    unsigned char gtSimdBaseJitType; // SIMD vector base JIT type
-    unsigned char gtSimdSize;        // SIMD vector size in bytes, use 0 for scalar intrinsics
+    template <unsigned InlineOperandCount>
+    GenTreeMultiOp(genTreeOps    oper,
+                   var_types     type,
+                   CompAllocator allocator,
+                   GenTree**     operands,
+                   size_t        operandCount,
+                   GenTree* (&inlineOperands)[InlineOperandCount] DEBUGARG(bool largeNode))
+        : GenTree(oper, type DEBUGARG(largeNode))
+    {
+        InitializeOperands(allocator, operands, operandCount, inlineOperands, InlineOperandCount);
+    }
 
 public:
+#if DEBUGGABLE_GENTREE
+    GenTreeMultiOp() : GenTree()
+    {
+    }
+#endif
+
+    GenTree*& Op(size_t index)
```

Comment:
I feel that these should be
```c++
GenTree* GetOp(size_t index) const;
void SetOp(size_t index, GenTree* value);
```

The assignment like
```c++
tree->Op(1) = op1;
```
take at least couple seconds from me to parse.

@dotnet/jit-contrib Anyone has the same opinion?

Reply:
I'm fine with both. We mostly use Get/Set pattern, but it's more verbose, and we already use rhs assignments, e.g.
![image](https://user-images.githubusercontent.com/523221/140606709-f935f490-981b-4968-a518-0e26ec78a119.png)

Reply:
FWIW, the reason I made these short is that I find the pattern of "`op` maintainance" (see [here](https://github.com/dotnet/runtime/blob/6547b55d2ac61db14a603f9b83eec7e9a353321e/src/coreclr/jit/morph.cpp#L12278-L12289) for an example) displeasing, and bug-prone (I spent at least a few hours on fixing one occurrence), so I would like to encourage people to use the accessors more liberally and let the native compiler do the CSEs for us.

Because of this I wanted it to be `Op`, but then it could not be `GenTree*` because in our codebase this is the prevailing pattern:
```c++
TThing& Thing();
TThing GetThing() const;
SetThing(TThing thing);
```

But I have no strong attachment to this (or arguments for it), and will happily rewrite them as we decided is best.